File: performance/src/ros2_t2.cpp
#include <cstdio>
#include <chrono>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "ros2_msg/msg/test.hpp"
using namespace std::chrono_literals;

class T2 : public rclcpp::Node
{
private:
    rclcpp::Subscription<ros2_msg::msg::Test>::SharedPtr subscription;
    rclcpp::Publisher<ros2_msg::msg::Test>::SharedPtr publisher;
    rclcpp::TimerBase::SharedPtr timer;
    size_t msg_no;
    const std::string bulb_data = std::string(1048576, '#');

    static auto age_of(const rclcpp::Time& time) {
        auto curr_time_nano = rclcpp::Clock().now().nanoseconds();
        auto time_nano = time.nanoseconds();
        return curr_time_nano-time_nano;
    }

    static auto mk_test_msg(const int msg_no, const std::string& data) {
        ros2_msg::msg::Test msg;
        msg.msg_no = msg_no;
        msg.data = data;
        msg.time_stamp = rclcpp::Clock().now();
        return msg;
    }

public:
    T2(): Node("T2"), msg_no(0) {
        RCLCPP_INFO(this->get_logger(), "Starting CppNode T2:");

        subscription = this->create_subscription<ros2_msg::msg::Test>( "/T1T", 10,
            [this](ros2_msg::msg::Test::UniquePtr msg) {
                RCLCPP_INFO(this->get_logger(), "%d,%d,%d", msg->msg_no, msg->data.length(), age_of(msg->time_stamp));
            });

        publisher = this->create_publisher<ros2_msg::msg::Test>("/T2T", 10);

        auto timer_callback = [this]() -> void {
            auto msg = mk_test_msg(this->msg_no++, this->bulb_data);
            this->publisher->publish(msg);
        };
        timer = this->create_wall_timer(50ms, timer_callback);
    }
};


int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<T2>());
    rclcpp::shutdown();
    return 0;
}

File: performance/src/ros2_t2.py
import rclpy
from rclpy.node import Node
from builtin_interfaces.msg import Time
from rclpy.clock import Clock
from ros2_msg.msg import Test


def age_of(time: Time):
    curr_time_nano = Clock().now().to_msg().nanosec
    time_nano = time.nanosec
    return curr_time_nano-time_nano


def mk_test_msg(msg_no: int, data: str) -> Test:
    msg = Test()
    msg.msg_no = msg_no
    msg.data = data
    msg.time_stamp = Clock().now().to_msg()
    return msg


class T2(Node):
    def __init__(self):
        super().__init__('T2')
        self.subscription = self.create_subscription(Test, '/T1T', self.subscription_callback, 10)
        self.publisher = self.create_publisher(Test, '/T2T', 10)
        self.timer = self.create_timer(0.05, self.timer_callback)
        self.msg_no = 0
        self.bulb_data = ''.rjust(1048576, '#')
        self.get_logger().info('Starting PyNode T2:')

    def subscription_callback(self, msg: Test):
        self.get_logger().info('%d,%d,%d' % (msg.msg_no, len(msg.data), age_of(msg.time_stamp)))

    def timer_callback(self):
        msg = mk_test_msg(self.msg_no, self.bulb_data)
        self.publisher.publish(msg)
        self.msg_no += 1


def main(args=None):
    rclpy.init(args=args)
    t2 = T2()
    rclpy.spin(t2)
    t2.destroy_node()
    rclpy.shutdown()
